dora/src/cannon/codegen.rs

```diff
@@ -1601,10 +1601,7 @@ where
         assert_eq!(self.argument_stack.len() as u32, num);
         let arguments = self.argument_stack.drain(..).collect::<Vec<_>>();
         let self_register = arguments[0];
-
-        let bytecode_type_self = self.bytecode.register_type(self_register);
         let position = self.bytecode.offset_position(self.current_offset.to_u32());
-        assert_eq!(bytecode_type_self, BytecodeType::Ptr);
```

Comment:
With this intact, it fails in `float0.dora` and `double0.dora` with:

```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `Int64`,
 right: `Ptr`', dora/src/cannon/codegen.rs:1615:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5
RUN: target/debug/dora  --compiler=cannon tests/double0.dora
```

Offending method is `String::binary`.

Reply:
InvokeDirect* checks whether the first parameter is nil and bails out if it is. The type of this argument therefore needs to be ptr. I think BytecodeGenerator emits the wrong bytecode: it should use InvokeStatic* if self isn't a ptr (see https://github.com/dinfuehr/dora/blob/f2557473b3d31944e21faf69606c92435fbd5577/dora/src/bytecode/generator.rs#L808-L811)